### DISCUSSION TITLE
Get an Audio example to work with an STM32 + STM32CubeIDE

### DISCUSSION THREAD

**Discussion by atgdms** [hathach/tinyusb#819] [Q&A]
So, I saw the discussion #633, and I followed all the instructions and got it to compile. Great!

But when I added the code from the device/audio_test example and I plugged it in, windows says Error Code 10 (This device cannot start). On Wireshark, it did not enumerate succesfully (comparing with a commercial microphone, see images)

What steps should I take ? Other than the steps provided in #633, I did the following :

- Added variables to main.c (from the main.c file in the example)
- Added fuctions to main.c 
- Copied startup function

My tusb_config.h, main.c and stm32l4xx.it are attached 
Working on a custom board with an STM32L4R9ZI.

Thanks in advance 
[code.zip](https://github.com/hathach/tinyusb/files/6443449/code.zip)

- TinyUSB enumeration
![wireshark_tinyusb](https://user-images.githubusercontent.com/78748955/117489997-89514480-af6e-11eb-93c5-6953e6bedd06.png)
- Commercial microphone enumeration
![wireshark_usbmic](https://user-images.githubusercontent.com/78748955/117490003-8a827180-af6e-11eb-888f-35806f2fc7fd.png)

**Comment by hathach** [comment]
L4 is not tested/supported yet https://github.com/hathach/tinyusb/issues/126 . There is some hipcup with the L4 probably due to crystal-less and or clock config

**Reply by atgdms** [reply]
Still, I am at least able to start the enumeration (as in #126) if it were a problem with some hardware (e.g. the clocks) it would not do that, right?
I will try to configure the clocks, as my custom board has one. If it works I will post it here

**Reply by atgdms** [reply]
Well, tested it and... same thing

So it is not just a matter of having an external crystal

**Reply by umichxuwentao** [reply]
Hi, have you been able to tackle this problem? I am having the exact same problem. I am working on a L476RG and a L4R5ZIP.

**Reply by atgdms** [reply]
Sorry, I wasn't able to. 
It has been a while since I worked on this project and finally I ended up leaving it aside. To solve the problem, a guy from my company rewrote the ST USB drivers, and made it work (I would love to share the code, but it is proprietary AND I didn't write it).

If you want to, I could try to retrace my steps and try to work with it together. I feel more mature with microcontrollers, and I feel more optimistic in trying to make it work.

But if want to try to make it work quickly, Azure RTOS seems to be a good choice (to be confirmed)

**Reply by umichxuwentao** [reply]
I have solved this problem. The way I solved it is to change the `tusb_config.h`  macro definition. I believe that there were some strange issues when I defined `CFG_TUSB_RHPORT0_MODE`. We cannot set it to be `(OPT_MODE_DEVICE | BOARD_DEVICE_FULL_SPEED)`, It should be `(OPT_MODE_DEVICE | BOARD_DEVICE_RHPORT_SPEED)`.

**Reply by umichxuwentao** [reply]
By the way, I was also trying to work from the ST USB drivers but did not have the time/energy to rewrote them. I just modified the USB descriptors and was stuck in the `DataIn` callback. I never managed to start the isochronous transfer just like the wrong `tusb-config.h` configuration. Though TinyUSB works now, I am still eager to know why isochronous transfer was stuck. Could you give me some hints about how to tackle this problem? Thanks a lot!
